src/web/mod.rs

```diff
@@ -310,33 +332,56 @@ fn match_version(conn: &Connection, name: &str, version: Option<&str>) -> MatchV
         .map(|v| if v == "newest" || v == "latest" { "*".to_owned() } else { v })
         .unwrap_or("*".to_string());
 
-    let versions = {
+    let (versions, corrected_name) = {
         let mut versions = Vec::new();
-        let rows = conn.query("SELECT versions FROM crates WHERE name = $1", &[&name]).unwrap();
-        if rows.len() == 0 {
-            return MatchVersion::None;
+        let mut rows = conn.query("SELECT versions FROM crates WHERE name = $1", &[&name]).unwrap();
+        let name_replace = if rows.len() == 0 {
+            // try looking up again with dashes and underscores replaced
+            if name.contains('-') {
```

Comment:
Hmm ... are we sure no crate will ever have both dashes and underscores?

Reply:
I know I have written one in the past, though I never published it. Should be possible to grep the index to check, I’ll try and take a look if I remember tomorrow.

Reply:
Finally got round to checking, there are 292 existing crates with a combination of `-` and `_` (found with `fd -- '-.*_|_.*-'` in a checkout of the index). They largely seem to be `-sys` crates, so unlikely to be manually typed in (one random interesting one: [`dash-and_underscore`](https://docs.rs/dash-and_underscore)).

<details><summary>click for list</summary>

```
a42-__-
advancedresearch-asi_core0
advancedresearch-graph_builder
advancedresearch-higher_order_core
advancedresearch-higher_order_point
advancedresearch-tree_mem_sort
advancedresearch-utility_programming
aflak_imgui-glium-renderer
aflak_imgui-sys
allegro_acodec-sys
allegro_audio-sys
allegro_color-sys
allegro_dialog-sys
allegro_font-sys
allegro_image-sys
allegro_memfile-sys
allegro_primitives-sys
allegro_ttf-sys
android_liblog-sys
android_log-sys
android_looper-sys
android_sensor-sys
aux_ulib-sys
aws-codebuild-status_aws
aws-codebuild-status_derive
aws-codebuild-status_server
aws-codebuild-status_terminal
aws-codebuild-status_web
blake2-rfc_bellman_edition
blip_buf-sys
bls_sigs_ref-rs
cargo-wasi-exe-x86_64-apple-darwin
cargo-wasi-exe-x86_64-pc-windows-msvc
cargo-wasi-exe-x86_64-unknown-linux-musl
cargo_crates-io_docs-rs_test
cargo_crates-io_docs-rs_test2
core_rustc-serialize
d-/_-
d-_-b
d3d10_1-sys
dash-and_underscore
deno-x86_64-apple-darwin
deno-x86_64-pc-windows-msvc
deno-x86_64-unknown-linux-gnu
dscfg-cached_file_storage
erlang_nif-sys
evr_vista-sys
exonum_librocksdb-sys
exonum_libsodium-sys
flexible-locks_derive
function_name-proc-macro
future-parking_lot
gdnative-visual_script
get_if_addrs-sys
gluon_c-api
gluon_language-server
google-accessapproval1_beta1
google-accessapproval1_beta1-cli
google-accesscontextmanager1_beta
google-accesscontextmanager1_beta-cli
google-adexchangebuyer2_v2_beta1
google-adexchangebuyer2_v2_beta1-cli
google-admin1_directory
google-admin1_directory-cli
google-admin1_reports
google-admin1_reports-cli
google-admin2_email_migration
google-alertcenter1_beta1
google-alertcenter1_beta1-cli
google-appengine1_beta4
google-appengine1_beta4-cli
google-appengine1_beta5
google-appengine1_beta5-cli
google-autoscaler1_beta2
google-autoscaler1_beta2-cli
google-binaryauthorization1_beta1
google-binaryauthorization1_beta1-cli
google-cloudasset1_beta1
google-cloudasset1_beta1-cli
google-clouderrorreporting1_beta1
google-clouderrorreporting1_beta1-cli
google-cloudkms1_beta1
google-cloudkms1_beta1-cli
google-cloudmonitoring2_beta2
google-cloudmonitoring2_beta2-cli
google-cloudprivatecatalog1_beta1
google-cloudprivatecatalog1_beta1-cli
google-cloudprivatecatalogproducer1_beta1
google-cloudprivatecatalogproducer1_beta1-cli
google-cloudresourcemanager1_beta1
google-cloudresourcemanager1_beta1-cli
google-cloudscheduler1_beta1
google-cloudscheduler1_beta1-cli
google-cloudtasks2_beta2
google-cloudtasks2_beta2-cli
google-cloudtasks2_beta3
google-cloudtasks2_beta3-cli
google-clouduseraccountsvm_beta
google-clouduseraccountsvm_beta-cli
google-commentanalyzer1_alpha1
google-commentanalyzer1_alpha1-cli
google-container1_beta1
google-containeranalysis1_beta1
google-containeranalysis1_beta1-cli
google-content2_sandbox
google-content2_sandbox-cli
google-dataflow1_b4
google-datafusion1_beta1
google-datafusion1_beta1-cli
google-datastore1_beta2
google-datastore1_beta3
google-datastore1_beta3-cli
google-deploymentmanager2_beta1
google-deploymentmanager2_beta2
google-deploymentmanager2_beta2-cli
google-dialogflow2_beta1
google-dialogflow2_beta1-cli
google-dlp2_beta1
google-dlp2_beta1-cli
google-dns1_beta1
google-factchecktools1_alpha1
google-factchecktools1_alpha1-cli
google-file1_beta1
google-file1_beta1-cli
google-firebase1_beta1
google-firebase1_beta1-cli
google-firebasehosting1_beta1
google-firebasehosting1_beta1-cli
google-firestore1_beta1
google-firestore1_beta1-cli
google-freebase1_sandbox
google-gamesconfiguration1_configuration
google-gamesconfiguration1_configuration-cli
google-gamesmanagement1_management
google-gamesmanagement1_management-cli
google-gan1_beta1
google-gan1_beta1-cli
google-genomics1_beta2
google-healthcare1_beta1
google-healthcare1_beta1-cli
google-iap1_beta1
google-iap1_beta1-cli
google-language1_beta1
google-language1_beta1-cli
google-logging1_beta3
google-logging2_beta1
google-logging2_beta1-cli
google-manager1_beta2
google-manager1_beta2-cli
google-ml1_beta1
google-ml1_beta1-cli
google-oauth2_v2
google-oslogin1_beta
google-oslogin1_beta-cli
google-proximitybeacon1_beta1
google-proximitybeacon1_beta1-cli
google-pubsub1_beta2
google-pubsub1_beta2-cli
google-replicapool1_beta2
google-replicapool1_beta2-cli
google-replicapoolupdater1_beta1
google-replicapoolupdater1_beta1-cli
google-reseller1_sandbox
google-reseller1_sandbox-cli
google-resourceviews1_beta2
google-resourceviews1_beta2-cli
google-runtimeconfig1_beta1
google-runtimeconfig1_beta1-cli
google-spectrum1_explorer
google-spectrum1_explorer-cli
google-speech1_beta1
google-speech1_beta1-cli
google-sqladmin1_beta4
google-sqladmin1_beta4-cli
google-taskqueue1_beta2
google-taskqueue1_beta2-cli
google-tpu1_alpha1
google-tpu1_alpha1-cli
google-videointelligence1_beta1
google-videointelligence1_beta1-cli
graphannis-malloc_size_of
graphannis-malloc_size_of_derive
gt-bijective_connection_graph
guessing_game-79888c11-a5a6-4548-a26a-c85ea39d0499
istherepie-hello_world
jh-x86_64
lettre-openssl111_email
libpg_query-sys
libzfs_core-sys
limn-text_layout
linux_input-sys
lpc177x_8x-hal
macro_rules_attribute-proc_macro
mbnapi_uuid-sys
mf_vista-sys
mfplat_vista-sys
mincore_downlevel-sys
minimult_cortex-m
msiz_rustc-ap-arena
msiz_rustc-ap-graphviz
msiz_rustc-ap-rustc_data_structures
msiz_rustc-ap-rustc_errors
msiz_rustc-ap-rustc_lexer
msiz_rustc-ap-rustc_macros
msiz_rustc-ap-rustc_target
msiz_rustc-ap-serialize
msiz_rustc-ap-syntax
msiz_rustc-ap-syntax_pos
msv1_0-sys
my_crate_72a9eb54-5002-4540-a6e6-f2620f51b550
next-gen_proc-macro
nix-ptsname_r-shim
noise_search_deps_librocksdb-sys
ntstc_libcmt-sys
ntstc_msvcrt-sys
optimisticpeach-opengles_graphics
paho_mqtt_c-sys
piston-ai_behavior
piston-button_controller
piston-button_tracker
piston-dyon_interactive
piston-fake_dpi
piston-gfx_texture
piston-graphics_api_version
piston-history_tree
piston-shaders_graphics2d
piston-split_controller
piston-timer_controller
piston2d-deform_grid
piston2d-drag_controller
piston2d-gfx_graphics
piston2d-glium_graphics
piston2d-graphics_tree
piston2d-opengl_graphics
piston2d-scroll_controller
piston2d-touch_visualizer
piston3d-gfx_voxel
pistoncore-event_loop
pistoncore-glfw_window
pistoncore-glutin_window
pistoncore-sdl2_window
pistoncore-winit_window
pocket_prover-derive
pocket_prover-set
quine-mc_cluskey
r2d2-influx_db_client
rest-client_codegen
rsudo-rs_support
rumblebars-rustlex_codegen
rust-riemann_health
rust_sodium-sys
rustc-ap-proc_macro
rustc-ap-rustc_cratesio_shim
rustc-ap-rustc_data_structures
rustc-ap-rustc_error_codes
rustc-ap-rustc_errors
rustc-ap-rustc_feature
rustc-ap-rustc_fs_util
rustc-ap-rustc_index
rustc-ap-rustc_lexer
rustc-ap-rustc_macros
rustc-ap-rustc_session
rustc-ap-rustc_target
rustc-ap-syntax_pos
rustdt-json_rpc
rustfmt-config_proc_macro
rv32m1_ri5cy-hal
rv32m1_ri5cy-pac
sapling-crypto_ce
screenruster-saver-laughing_man
serde_sane-rs
solana_libra_crypto-derive
solana_libra_grpcio-client
struct_deser-derive
sudo_plugin-sys
toml-query_derive
twain_32-sys
unic-ucd-name_aliases
va_list-helper
va_list-rs
va_list-test
vds_uuid-sys
vss_uuid-sys
web-view_suppress
webview-sys_suppress
winapi-x86_64-pc-windows-gnu
wlc-sys-with_elogind
wlc-with_elogind
ws2_32-sys
wsbapp_uuid-sys
x86_64-xsave
xinput9_1_0-sys
```

</details>